cfft/src/radix_2_butterfly.rs

```diff
@@ -0,0 +1,278 @@
+            let block_size = 1 << (i + 1);
+            let half_block_size = block_size >> 1;
+
+            for chuncks in mat.values.chunks_mut(block_size * width) {
+                let (low_chunks, high_chunks) = chuncks.split_at_mut(half_block_size * width);
+
+                low_chunks
+                    .chunks_mut(width)
+                    .zip(high_chunks.chunks_mut(width).rev())
+                    .zip(twiddle)
+                    .for_each(|((low_chunk, hi_chunk), twiddle)| {
+                        butterfly_icfft(low_chunk, hi_chunk, *twiddle)
+                    })
+            }
+        }
+
+        mat
+    }
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////
+//////////////////////////////////////////////////////////////////////////////////////////////
+//////////////////////////////////////////////////////////////////////////////////////////////
+
+fn butterfly_cfft<Base: Field>(low_chunk: &mut [Base], high_chunk: &mut [Base], twiddle: Base) {
+    for (low, high) in low_chunk.iter_mut().zip(high_chunk) {
+        let sum = *low + *high;
+        let diff = (*low - *high) * twiddle;
+        *low = sum;
+        *high = diff;
+    }
+}
+
+fn butterfly_icfft<Base: Field>(low_chunk: &mut [Base], high_chunk: &mut [Base], twiddle: Base) {
+    for (low, high) in low_chunk.iter_mut().zip(high_chunk) {
+        let high_twiddle = *high * twiddle;
+        let sum = *low + high_twiddle;
+        let diff = *low - high_twiddle;
+        *low = sum;
+        *high = diff;
+    }
+}
+
+// Code for computing the twiddles.
+// Currently, this requires us to recompute the twiddles every time we want to use the CFFT.
+// This is somewhat expensive so, in the long run, it would be ideal to modify the structure so they can be precomputed.
+
+/// Compute the twiddles for the CFFT.
+/// Let N = 2^n be the size of our initial set. Then we start with the domain
+/// {g, g^3, ..., g^{-3}, g^{-1}} for g a 2N'th root of unity.
+/// The initial twiddle domain is the first half of the full domain.
+/// In the first step our twiddles are the inverse imaginary parts and we simply halve the domain size.
+/// In all subsequent steps our twiddles are the inverse real parts and we both halve the domain size and square every element.
+pub fn cfft_twiddles<Base: Field, Ext: ComplexExtension<Base>>(log_n: usize) -> Vec<Vec<Base>> {
+    let size = 1 << (log_n - 1);
+    let init_domain = cfft_domain::<Base, Ext>(log_n, size); // Get the starting domain.
+
+    let mut working_domain: Vec<_> = init_domain
+        .iter()
+        .take(size / 2)
+        .map(|x| x.real())
+        .collect(); // After the first step we only need the real part.
+
+    (0..log_n)
+        .map(|i| {
+            let size = working_domain.len();
+            if i == 0 {
+                init_domain.iter().map(|x| x.imag().inverse()).collect() // The twiddles in step one are the inverse imaginary parts.
```

Comment:
Consider using `batch_multiplicative_inverse` here. That's quite a bit faster, and it might also make it easier to merge with `cfft_inv_twiddles`: take an argument `inv: bool`, and
```rust
if inv {
  batch_mutiplicative_inverse(output)
}
```

Reply:
I've added this but currently just done it in the naive way which applies batch_mutiplicative_inverse(&mut output) inside the loop.

Long term it would be faster to apply this outside the loop so we only need to use it once but that will involve a larger restructure so I'm leaving it for now.